src/xmpp/xmpp-im/xmpp_tasks.cpp

```diff
@@ -860,12 +860,13 @@ static QDomElement oldStyleNS(const QDomElement &e)
     return i;
 }
 
-JT_Message::JT_Message(Task *parent, const Message &msg)
+JT_Message::JT_Message(Task *parent, Message &msg, EncryptionHandler *encryptionHandler)
 :Task(parent)
 {
-    m = msg;
-    if (m.id().isEmpty())
-        m.setId(id());
+    m = &msg;
```

Comment:
This doesn't look good. It explicitly forbids any optimizations of snding workflow.
If you really want to have CPU optimization on copy, try any of copy-on-write classes or even one of Explicitly shared classes.

Reply:
I'm not sure I understand what sending workflow optimization you mean, but switching to a reference was necessary here not to get rid of the copying. It's to allow the encryption handler to set the `wasEncrypted` flag, so this fact can be reflected in the UI after the message has been sent.

Reply:
Basically, we can build priority queues, what means the message won't be send immediately and then we  get dangling pointer.

Reply:
Please see the updated implementation. Now instead of storing a `Message` in `JT_Message` we build `QDomNode` from it right in `JT_Message` constructor, encrypt it if necessary and then store the resulting `QDomNode`.

Reply:
Ok